man/man1/pquery.1
.\" Automatically generated by Pandoc 2.14.2
.\"
.TH "pquery" "1" "" "2021-08-17" "Open PMIx"
.hy
.SH NAME
.PP
pquery - Request information from the system
.SH SYNOPSIS
.IP
.nf
\f[C]
pquery [ options ] [ keys ]
\f[R]
.fi
.SH DESCRIPTION
.PP
\f[C]pquery\f[R] will request and output the specified information.
This requires that \f[C]pquery\f[R] connect to an appropriate server -
the tool will automatically attempt to do so, but options are provided
to assist its search or to direct it to a specific server.
.PP
Keys passed to \f[C]pquery\f[R] may optionally include one or more
qualifiers, with the individual qualifiers delimited by semi-colons.
For example:
.IP
.nf
\f[C]
PMIX_STORAGE_XFER_SIZE[PMIX_STORAGE_ID=lustre1,lustre2;PMIX_STORAGE_PATH=foo]
\f[R]
.fi
.SH OPTIONS
.PP
\f[C]pquery\f[R] supports the following options:
.IP
.nf
\f[C]
-h|--help                This help message
-n|--nspace <arg0>       Specify server nspace to connect to
-p|--pid <arg0>          Specify server pid to connect to
   --system-server       Specifically connect to the system server
   --system-server-first Look for the system server first
   --uri <arg0>          Specify URI of server to connect to
-v|--verbose             Be Verbose
\f[R]
.fi
.SH AUTHORS
.PP
The OpenPMIx maintainers \[en] see https://github.com/openpmix/openpmix
or the file \f[C]AUTHORS\f[R].


NEWS
Copyright (c) 2015-2020 Intel, Inc.  All rights reserved.
Copyright (c) 2017-2020 IBM Corporation.  All rights reserved.
Copyright (c) 2021      Nanook Consulting.  All rights reserved.
Copyright (c) 2021      Amazon.com, Inc. or its affiliates.
                        All Rights reserved.
$COPYRIGHT$

Additional copyrights may follow

$HEADER$

===========================================================================

This file contains the main features as well as overviews of specific
bug fixes (and other actions) for each version of PMIx since
version 1.0.

As more fully described in the "Software Version Number" section in
the README file, PMIx typically maintains two separate version
series simultaneously - the current release and one that is locked
to only bug fixes.  Since these series are semi-independent of each
other, a single NEWS-worthy item might apply to different series. For
example, a bug might be fixed in the master, and then moved to
multiple release branches.

4.1.1 -- TBD
----------------------
**** NOTE: As of v4.1.1, PMIx includes an EXCEPTIONS file that lists
**** all deviations from the PMIx Standard. This primarily includes
**** extensions that have not yet been adopted by the Standard.

**** NOTE: As of v4.1.1, PMIx no longer has a dependency on "pandoc"
**** for building and installing man pages.

 - PR #2299: Update 4.1.1 rc2
    - Update VERSION and NEWS
    - Add some spawn-specific timeout attributes
    - Resolve race condition in lost connection
    - Provide "partial_success" error when collectives not complete
    - Only conditionally decode the nspace return value when we
        are sure the spawn was successful
    - Correctly copy stone age hwloc topologies
    - Remove man page Markdown source and build dependency on pandoc
 - PR #2277: Update v4.1.1 release candidate
    - Add missing m4 file
    - Add a Standard extension value to the compliance version
    - Properly read/output stdout/err from a fork/exec'd child
    - Default to using our local_output flag
    - Cleanup compiler warnings for ancient hwloc versions
    - Prefix the output files with "pmix"
    - Ensure tools wait until all active events are processed
    - autogen.pl: ignore all excluded components
    - Don't treat inability to open shmem file as fatal
    - Avoid use of MCA params for singleton and report-uri
    - Ensure the server waits for all IOF and message events to complete
    - Restore the thread join in progress thread "stop"
    - Mark the read event as no longer active
    - Avoid blocking in the stdin read handler
    - Some cleanup of IOF output
    - Add missing .m4 files to extra_dist
    - Check for libevent minimum 2.0.21
    - Add Intel GPU component
    - Correct vendor IDs and generalize check_vendor
    - Add missing storage-related datatype support
    - Add missing storage constants
    - Improve pnet component selection
    - Cleanup the device distance computation


4.1.0 -- 29 July 2021
----------------------
**** NOTE: This release implements the complete PMIx v4.1 Standard
**** and therefore includes a number of new APIs and features. These
**** are fully documented in the official document. It also includes
**** some extensions that have not yet been included in that document.

Beyond the v4.1 modifications and additions to APIs, datatypes, attributes,
and macros, changes to the library include:

 - PR #2251: More updates from master
     - Replaced PMIx_Notify_Event with cbfunc call in errhandler to match
     - Update attribute support tables
 - PR #2248: Continue updates to support MPICH integration
     - Extend IOF outputting format to cover Hydra options.
 - PR #2246: Cleanup some IOF attributes
 - PR #2235: Cherry-pick updates from master branch
     - Default tools to outputting their IOF
     - Initalize val before get in case get isn't successful
     - Some cleanups of the event notification and keepalive support
     - Remove stale travis.yml file
     - Update simptest to truly support PMIx_Abort
     - Some cleanups for client finalize and IOF output
     - Do not forward cached IOF to self
     - Update how C to Python bytes/strings are handled for get and byte objects
     - Add test_v2 directory
     - Add configure logic for RM and sse support
     - Add pstat framework
     - Remove duplicate PMIx_Data_load and PMIx_Data_unload definitions
     - Add manpage files for tools
     - Add sse common component
     - Add the prm components
     - Add the storage framework
     - General update of code base to track master branch
 - PR #2224: common/dstore: Fix inconsistent Makefile.am
 - PR #2216: Cherry-pick updates from master branch
     - Add missing osname endpt elements to bfrops
     - Optimize check for nodes
     - Transfer stdout/err formatting to PMIx
     - Ensure tool output of IOF
 - PR #2208: Cherry-pick updates from master branch
     - Enable re-init of clients
     - Add attribute to indicate copy/nocopy of output directed to files
 - PR #2204: Add a few job error constants
 - PR #2201: Cherry-pick updates from master branch
     - Minor updates based on Standards review
     - Correct references to help-ptl-tool.txt
     - Protect register_nspace against new entries
     - Add oversubscribed attribute
 - PR #2195: configury: Use AC_CHECK_ALIGNOF and fix cross-compiling
 - PR #2190:Cherry-pick updates from master branch
     - Remove duplicative pmdl/ompi directories
     - Add missing m4 file

src/tools/pquery/pquery.c
/*
 * Copyright (c) 2004-2010 The Trustees of Indiana University and Indiana
 *                         University Research and Technology
 *                         Corporation.  All rights reserved.
 * Copyright (c) 2004-2011 The University of Tennessee and The University
 *                         of Tennessee Research Foundation.  All rights
 *                         reserved.
 * Copyright (c) 2004-2005 High Performance Computing Center Stuttgart,
 *                         University of Stuttgart.  All rights reserved.
 * Copyright (c) 2004-2005 The Regents of the University of California.
 *                         All rights reserved.
 * Copyright (c) 2006-2013 Los Alamos National Security, LLC.
 *                         All rights reserved.
 * Copyright (c) 2009-2012 Cisco Systems, Inc.  All rights reserved.
 * Copyright (c) 2011      Oak Ridge National Labs.  All rights reserved.
 * Copyright (c) 2013-2020 Intel, Inc.  All rights reserved.
 * Copyright (c) 2015      Mellanox Technologies, Inc.  All rights reserved.
 * Copyright (c) 2021      Nanook Consulting.  All rights reserved.
 * $COPYRIGHT$
 *
 * Additional copyrights may follow
 *
 * $HEADER$
 *
 */

#include "pmix_config.h"
#include "include/pmix.h"
#include "include/pmix_common.h"
#include "include/pmix_server.h"

#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>

#include "include/pmix_tool.h"
#include "src/common/pmix_attributes.h"
#include "src/mca/base/base.h"
#include "src/mca/pinstalldirs/base/base.h"
#include "src/runtime/pmix_rte.h"
#include "src/threads/threads.h"
#include "src/util/cmd_line.h"
#include "src/util/keyval_parse.h"
#include "src/util/show_help.h"

typedef struct {
    pmix_lock_t lock;
    pmix_status_t status;
} mylock_t;

static pmix_proc_t myproc;

/* define a structure for collecting returned
 * info from a query */
typedef struct {
    pmix_lock_t lock;
    pmix_status_t status;
    pmix_info_t *info;
    size_t ninfo;
} myquery_data_t;

/* this is a callback function for the PMIx_Query
 * API. The query will callback with a status indicating
 * if the request could be fully satisfied, partially
 * satisfied, or completely failed. The info parameter
 * contains an array of the returned data, with the
 * info->key field being the key that was provided in
 * the query call. Thus, you can correlate the returned
 * data in the info->value field to the requested key.
 *
 * Once we have dealt with the returned data, we must
 * call the release_fn so that the PMIx library can
 * cleanup */
static void querycbfunc(pmix_status_t status, pmix_info_t *info, size_t ninfo, void *cbdata,
                        pmix_release_cbfunc_t release_fn, void *release_cbdata)
{
    myquery_data_t *mq = (myquery_data_t *) cbdata;
    size_t n;

    mq->status = status;
    /* save the returned info - the PMIx library "owns" it
     * and will release it and perform other cleanup actions
     * when release_fn is called */
    if (0 < ninfo) {
        PMIX_INFO_CREATE(mq->info, ninfo);
        mq->ninfo = ninfo;
        for (n = 0; n < ninfo; n++) {
            PMIX_INFO_XFER(&mq->info[n], &info[n]);
        }
    }

    /* let the library release the data and cleanup from
     * the operation */
    if (NULL != release_fn) {
        release_fn(release_cbdata);
    }

    /* release the block */
    PMIX_WAKEUP_THREAD(&mq->lock);
}

/* this is the event notification function we pass down below
 * when registering for general events - i.e.,, the default
 * handler. We don't technically need to register one, but it
 * is usually good practice to catch any events that occur */
static void notification_fn(size_t evhdlr_registration_id, pmix_status_t status,
                            const pmix_proc_t *source, pmix_info_t info[], size_t ninfo,
                            pmix_info_t results[], size_t nresults,
                            pmix_event_notification_cbfunc_fn_t cbfunc, void *cbdata)
{
    /* this example doesn't do anything with default events */
    if (NULL != cbfunc) {
        cbfunc(PMIX_EVENT_ACTION_COMPLETE, NULL, 0, NULL, NULL, cbdata);
    }
}

/* event handler registration is done asynchronously because it
 * may involve the PMIx server registering with the host RM for
 * external events. So we provide a callback function that returns
 * the status of the request (success or an error), plus a numerical index
 * to the registered event. The index is used later on to deregister
 * an event handler - if we don't explicitly deregister it, then the
 * PMIx server will do so when it see us exit */
static void evhandler_reg_callbk(pmix_status_t status, size_t evhandler_ref, void *cbdata)
{
    mylock_t *lock = (mylock_t *) cbdata;

    if (PMIX_SUCCESS != status) {
        fprintf(stderr, "Client %s:%d EVENT HANDLER REGISTRATION FAILED WITH STATUS %d, ref=%lu\n",
                myproc.nspace, myproc.rank, status, (unsigned long) evhandler_ref);
    }
    lock->status = status;
    PMIX_WAKEUP_THREAD(&lock->lock);
}

/*****************************************
 * Global Vars for Command line Arguments
 *****************************************/
typedef struct {
    bool help;
    bool verbose;
    pid_t pid;
    char *nspace;
    char *uri;
    bool sysfirst;
    bool system;
    char *client;
    char *server;
    char *tool;
    char *host;
    bool clientfns;
    bool serverfns;
    bool toolfns;
    bool hostfns;
} pmix_pquery_globals_t;

pmix_pquery_globals_t pmix_pquery_globals = {0};

pmix_cmd_line_init_t cmd_line_opts[]
    = {{NULL, 'h', NULL, "help", 0, &pmix_pquery_globals.help, PMIX_CMD_LINE_TYPE_BOOL,
        "This help message"},

       {NULL, 'v', NULL, "verbose", 0, &pmix_pquery_globals.verbose, PMIX_CMD_LINE_TYPE_BOOL,
        "Be Verbose"},

       {NULL, 'p', NULL, "pid", 1, &pmix_pquery_globals.pid, PMIX_CMD_LINE_TYPE_INT,
        "Specify server pid to connect to"},

       {NULL, 'n', NULL, "nspace", 1, &pmix_pquery_globals.nspace, PMIX_CMD_LINE_TYPE_STRING,
        "Specify server nspace to connect to"},

       {NULL, '\0', NULL, "uri", 1, &pmix_pquery_globals.uri, PMIX_CMD_LINE_TYPE_STRING,
        "Specify URI of server to connect to"},

       {NULL, '\0', NULL, "system-server-first", 0, &pmix_pquery_globals.sysfirst,
        PMIX_CMD_LINE_TYPE_BOOL, "Look for the system server first"},

       {NULL, '\0', NULL, "system-server", 0, &pmix_pquery_globals.system, PMIX_CMD_LINE_TYPE_BOOL,
        "Specifically connect to the system server"},

       /* End of list */
       {NULL, '\0', NULL, NULL, 0, NULL, PMIX_CMD_LINE_TYPE_NULL, NULL}};

int main(int argc, char **argv)
{
    pmix_status_t rc;
    pmix_info_t *info;
    mylock_t mylock;
    pmix_cmd_line_t cmd_line;
    size_t n, m, k, nqueries, ninfo, ndarray;
    myquery_data_t mq = {{0}};
    int count;
    char **qkeys = NULL;
    const char *attr;
    bool server = false;
    pmix_list_t querylist, qlist;
    pmix_querylist_t *qry;
    char **qprs;
    char *strt, *endp, *kptr;
    pmix_infolist_t *iptr;
    char *str, *args = NULL;
    pmix_query_t *queries;
    pmix_info_t *infoptr, *ifptr;
    uint64_t u64;

    /* protect against problems if someone passes us thru a pipe
     * and then abnormally terminates the pipe early */
    signal(SIGPIPE, SIG_IGN);

    /* initialize the output system */
    if (!pmix_output_init()) {
        return PMIX_ERROR;
    }

    /* initialize install dirs code */
    if (PMIX_SUCCESS
        != (rc = pmix_mca_base_framework_open(&pmix_pinstalldirs_base_framework,
                                              PMIX_MCA_BASE_OPEN_DEFAULT))) {
        fprintf(stderr,
                "pmix_pinstalldirs_base_open() failed -- process will likely abort (%s:%d, "
                "returned %d instead of PMIX_SUCCESS)\n",
                __FILE__, __LINE__, rc);
        return rc;
    }
    if (PMIX_SUCCESS != (rc = pmix_pinstall_dirs_base_init(NULL, 0))) {
        fprintf(stderr,
                "pmix_pinstalldirs_base_init() failed -- process will likely abort (%s:%d, "
                "returned %d instead of PMIX_SUCCESS)\n",
                __FILE__, __LINE__, rc);
        return rc;
    }

    /* initialize the help system */
    pmix_show_help_init();

    /* keyval lex-based parser */
    if (PMIX_SUCCESS != (rc = pmix_util_keyval_parse_init())) {
        fprintf(stderr, "pmix_util_keyval_parse_init failed with %d\n", rc);
        return PMIX_ERROR;
    }

    /* Setup the parameter system */
    if (PMIX_SUCCESS != (rc = pmix_mca_base_var_init())) {
        fprintf(stderr, "pmix_mca_base_var_init failed with %d\n", rc);
        return PMIX_ERROR;
    }

    /* register params for pmix */
    if (PMIX_SUCCESS != (rc = pmix_register_params())) {
        fprintf(stderr, "pmix_register_params failed with %d\n", rc);
        return PMIX_ERROR;
    }

    /* Parse the command line options */
    pmix_cmd_line_create(&cmd_line, cmd_line_opts);

    rc = pmix_cmd_line_parse(&cmd_line, true, false, argc, argv);

    if (PMIX_SUCCESS != rc) {
        if (PMIX_ERR_SILENT != rc) {
            fprintf(stderr, "%s: command line error (%s)\n", argv[0], PMIx_Error_string(rc));
        }
        return rc;
    }

    if (pmix_pquery_globals.help) {
        args = pmix_cmd_line_get_usage_msg(&cmd_line);
        str = pmix_show_help_string("help-pquery.txt", "usage", true, args);
        if (NULL != str) {
            printf("%s", str);
            free(str);
        }
        free(args);
        /* If we show the help message, that should be all we do */
        exit(0);
    }

    /* get the argv array of keys they want us to query */
    pmix_cmd_line_get_tail(&cmd_line, &count, &qkeys);

    /* if they didn't give us an option, then we can't do anything */
    if (NULL == qkeys) {
        args = pmix_cmd_line_get_usage_msg(&cmd_line);
        str = pmix_show_help_string("help-pquery.txt", "usage", true, args);
        if (NULL != str) {
            printf("%s", str);
            free(str);
        }
        free(args);
        exit(1);
    }

    /* they might be querying the system about a client, server, or tool
     * attribute, so register those - this will allow us to compare the
     * provided key with the actual name of the attribute so we can
     * identify it */
    pmix_init_registered_attrs();
    pmix_register_client_attrs();
    pmix_register_server_attrs();
    pmix_register_tool_attrs();

    /* if we were given the pid of a starter, then direct that
     * we connect to it */
    n = 1;
    PMIX_INFO_CREATE(info, n);
    if (0 < pmix_pquery_globals.pid) {
        PMIX_INFO_LOAD(&info[0], PMIX_SERVER_PIDINFO, &pmix_pquery_globals.pid, PMIX_PID);
    } else if (NULL != pmix_pquery_globals.nspace) {
        PMIX_INFO_LOAD(&info[0], PMIX_SERVER_NSPACE, pmix_pquery_globals.nspace, PMIX_STRING);
    } else if (NULL != pmix_pquery_globals.uri) {
        PMIX_INFO_LOAD(&info[0], PMIX_SERVER_URI, pmix_pquery_globals.uri, PMIX_STRING);
    } else if (pmix_pquery_globals.sysfirst) {
        /* otherwise, use the system connection first, if available */
        PMIX_INFO_LOAD(&info[0], PMIX_CONNECT_SYSTEM_FIRST, NULL, PMIX_BOOL);
    } else if (pmix_pquery_globals.system) {
        PMIX_INFO_LOAD(&info[0], PMIX_CONNECT_TO_SYSTEM, NULL, PMIX_BOOL);
    } else {
        /* we set ourselves up as a server, but no connections required */
        PMIX_INFO_LOAD(&info[0], PMIX_GDS_MODULE, "hash", PMIX_STRING);
        server = true;
        ;
    }

    if (server) {
        /* initialize as a server so we can process the query ourselves */
        rc = PMIx_server_init(NULL, info, n);
        if (PMIX_SUCCESS != rc) {
            fprintf(stderr, "PMIx_server_init failed: %s\n", PMIx_Error_string(rc));
            exit(rc);
        }
    } else {
        /* init as a tool */
        rc = PMIx_tool_init(&myproc, info, n);
        if (PMIX_SUCCESS != rc) {
            fprintf(stderr, "PMIx_tool_init failed: %s\n", PMIx_Error_string(rc));
            exit(rc);
        }
        PMIX_INFO_FREE(info, 1);
    }

    /* register a default event handler */
    PMIX_CONSTRUCT_LOCK(&mylock.lock);
    PMIx_Register_event_handler(NULL, 0, NULL, 0, notification_fn, evhandler_reg_callbk,
                                (void *) &mylock);
    PMIX_WAIT_THREAD(&mylock.lock);
    if (PMIX_SUCCESS != mylock.status) {
        fprintf(stderr, "PMIx_Register_event_handler returned bad status: %d\n", mylock.status);
        PMIX_DESTRUCT_LOCK(&mylock.lock);
        rc = mylock.status;
        goto done;
    }
    PMIX_DESTRUCT_LOCK(&mylock.lock);

    /* generate the queries */
    PMIX_CONSTRUCT(&querylist, pmix_list_t);
    for (n = 0; NULL != qkeys[n]; n++) {
        qry = PMIX_NEW(pmix_querylist_t);
        PMIX_CONSTRUCT(&qlist, pmix_list_t);
        /* check for qualifiers: key[qual=foo;qual=bar] */
        if (NULL != (strt = strchr(qkeys[n], '['))) {
            /* we have qualifiers - find the end */
            *strt = '\0';
            ++strt;
            if (NULL == (endp = strrchr(strt, ']'))) {
                str = pmix_show_help_string("help-pquery.txt", "bad-quals", true, qkeys[n]);
                if (NULL != str) {
                    printf("%s", str);
                    free(str);
                }
                exit(1);
            }
            *endp = '\0';
            /* break into qual=val pairs */
            qprs = pmix_argv_split(strt, ';');
            for (m = 0; NULL != qprs[m]; m++) {
                /* break each pair */
                if (NULL == (kptr = strchr(qprs[m], '='))) {
                    str = pmix_show_help_string("help-pquery.txt", "bad-qual", true, qkeys[n],
                                                qprs[m]);
                    if (NULL != str) {
                        printf("%s", str);
                        free(str);
                    }
                    exit(1);
                }
                *kptr = '\0';
                kptr++;
                iptr = PMIX_NEW(pmix_infolist_t);
                if (NULL == (attr = pmix_attributes_lookup(qprs[m]))) {
                    exit(1);
                }
                PMIX_INFO_LOAD(&iptr->info, attr, kptr, PMIX_STRING);
                pmix_list_append(&qlist, &iptr->super);
            }
        }
        /* convert the key */
        if (NULL == (attr = pmix_attributes_lookup(qkeys[n]))) {
            exit(1);
        }
        pmix_argv_append_nosize(&qry->query.keys, attr);
        /* add in any qualifiers */
        m = pmix_list_get_size(&qlist);
        if (0 < m) {
            PMIX_QUERY_QUALIFIERS_CREATE(&qry->query, m);
            m = 0;
            PMIX_LIST_FOREACH (iptr, &qlist, pmix_infolist_t) {
                PMIX_INFO_XFER(&qry->query.qualifiers[m], &iptr->info);
                ++m;
            }
        }
        pmix_list_append(&querylist, &qry->super);
        PMIX_LIST_DESTRUCT(&qlist);
    }
    /* convert the list of queries into an array */
    nqueries = pmix_list_get_size(&querylist);
    PMIX_QUERY_CREATE(queries, nqueries);
    m = 0;
    PMIX_LIST_FOREACH (qry, &querylist, pmix_querylist_t) {
        /* move the queries across */
        queries[m].keys = qry->query.keys;
        queries[m].nqual = qry->query.nqual;
        queries[m].qualifiers = qry->query.qualifiers;
        ++m;
    }
    PMIX_LIST_DESTRUCT(&querylist);

    PMIX_CONSTRUCT_LOCK(&mq.lock);
    rc = PMIx_Query_info_nb(queries, nqueries, querycbfunc, (void *) &mq);
    if (PMIX_SUCCESS != rc) {
        fprintf(stderr, "PMIx_Query_info failed: %d\n", rc);
        goto done;
    }
    PMIX_WAIT_THREAD(&mq.lock);
    PMIX_DESTRUCT_LOCK(&mq.lock);
    if (PMIX_SUCCESS != mq.status) {
        fprintf(stderr, "PMIx_Query_info returned: %s\n", PMIx_Error_string(mq.status));
        rc = mq.status;
    } else {
        /* print out the returned value(s) */
        for (n = 0; n < mq.ninfo; n++) {
            if (NULL == (attr = pmix_attributes_reverse_lookup(mq.info[n].key))) {
                fprintf(stdout, "%s: ", mq.info[n].key);
            } else {
                fprintf(stdout, "%s: ", attr);
            }
            if (PMIX_STRING == mq.info[n].value.type) {
                fprintf(stdout, "%s\n", mq.info[n].value.data.string);
            } else if (PMIX_DATA_ARRAY == mq.info[n].value.type) {
                fprintf(stdout, "\n");
                infoptr = (pmix_info_t *) mq.info[n].value.data.darray->array;
                ninfo = mq.info[n].value.data.darray->size;
                for (m = 0; m < ninfo; m++) {
                    if (NULL == (attr = pmix_attributes_reverse_lookup(infoptr[m].key))) {
                        fprintf(stdout, "\t%s:", infoptr[m].key);
                    } else {
                        fprintf(stdout, "\t%s:", attr);
                    }
                    if (PMIX_STRING == infoptr[m].value.type) {
                        fprintf(stdout, "  %s\n", infoptr[m].value.data.string);
                    } else if (PMIX_DATA_ARRAY == infoptr[m].value.type) {
                        fprintf(stdout, "\n");
                        ifptr = (pmix_info_t *) infoptr[m].value.data.darray->array;
                        ndarray = infoptr[m].value.data.darray->size;
                        for (k = 0; k < ndarray; k++) {
                            if (NULL == (attr = pmix_attributes_reverse_lookup(ifptr[k].key))) {
                                fprintf(stdout, "\t\t%s:", ifptr[k].key);
                            } else {
                                fprintf(stdout, "\t\t%s:", attr);
                            }
                            if (PMIX_STRING == ifptr[k].value.type) {
                                fprintf(stdout, "  %s\n", ifptr[k].value.data.string);
                            } else if (PMIX_PROC_RANK == ifptr[k].value.type) {
                                fprintf(stdout, "  %u\n", ifptr[k].value.data.rank);
                            } else {
                                /* see if it is a number */
                                PMIX_VALUE_GET_NUMBER(rc, &ifptr[k].value, u64, uint64_t);
                                if (PMIX_SUCCESS == rc) {
                                    fprintf(stdout, "  %lu\n", (unsigned long) u64);
                                } else {
                                    fprintf(stdout, "  Unimplemented value type: %s\n",
                                            PMIx_Data_type_string(ifptr[k].value.type));
                                }
                            }
                        }
                    } else if (PMIX_PROC_RANK == infoptr[m].value.type) {
                        fprintf(stdout, "  %u\n", infoptr[m].value.data.rank);
                    } else {
                        /* see if it is a number */
                        PMIX_VALUE_GET_NUMBER(rc, &infoptr[m].value, u64, uint64_t);
                        if (PMIX_SUCCESS == rc) {
                            fprintf(stdout, "  %lu\n",
                                    (unsigned long) infoptr[m].value.data.uint64);
                        } else {
                            fprintf(stdout, "  Unimplemented value type: %s\n",
                                    PMIx_Data_type_string(infoptr[m].value.type));
                        }
                    }
                }
            }
        }
    }

done:
    if (server) {
        PMIx_server_finalize();
    } else {
        PMIx_tool_finalize();
    }

    return (rc);
}
